core/src/main/scala/org/apache/spark/NewAccumulator.scala

```diff
@@ -0,0 +1,356 @@
+
+import java.{lang => jl}
+import java.io.ObjectInputStream
+import java.util.concurrent.atomic.AtomicLong
+import javax.annotation.concurrent.GuardedBy
+
+import org.apache.spark.scheduler.AccumulableInfo
+import org.apache.spark.util.Utils
+
+
+private[spark] case class AccumulatorMetadata(
+    id: Long,
+    name: Option[String],
+    countFailedValues: Boolean) extends Serializable
+
+
+/**
+ * The base class for accumulators, that can accumulate inputs of type `IN`, and produce output of
+ * type `OUT`.  Implementations must define following methods:
+ *  - isZero:       tell if this accumulator is zero value or not. e.g. for a counter accumulator,
+ *                  0 is zero value; for a list accumulator, Nil is zero value.
+ *  - copyAndReset: create a new copy of this accumulator, which is zero value. i.e. call `isZero`
+ *                  on the copy must return true.
+ *  - add:          defines how to accumulate the inputs. e.g. it can be a simple `+=` for counter
+ *                  accumulator
+ *  - merge:        defines how to merge another accumulator of same type.
+ *  - localValue:   defines how to produce the output by the current state of this accumulator.
+ *
+ * The implementations decide how to store intermediate values, e.g. a long field for a counter
+ * accumulator, a double and a long field for a average accumulator(storing the sum and count).
+ */
+abstract class NewAccumulator[IN, OUT] extends Serializable {
+  private[spark] var metadata: AccumulatorMetadata = _
+  private[this] var atDriverSide = true
+
+  private[spark] def register(
+      sc: SparkContext,
+      name: Option[String] = None,
+      countFailedValues: Boolean = false): Unit = {
+    if (this.metadata != null) {
+      throw new IllegalStateException("Cannot register an Accumulator twice.")
+    }
+    this.metadata = AccumulatorMetadata(AccumulatorContext.newId(), name, countFailedValues)
+    AccumulatorContext.register(this)
+    sc.cleaner.foreach(_.registerAccumulatorForCleanup(this))
+  }
+
+  final def isRegistered: Boolean =
+    metadata != null && AccumulatorContext.originals.containsKey(metadata.id)
+
+  private def assertMetadataNotNull(): Unit = {
+    if (metadata == null) {
+      throw new IllegalAccessError("The metadata of this accumulator has not been assigned yet.")
+    }
+  }
+
+  final def id: Long = {
+    assertMetadataNotNull()
+    metadata.id
+  }
+
+  final def name: Option[String] = {
+    assertMetadataNotNull()
+    metadata.name
+  }
+
+  final def countFailedValues: Boolean = {
+    assertMetadataNotNull()
+    metadata.countFailedValues
+  }
+
+  private[spark] def toInfo(update: Option[Any], value: Option[Any]): AccumulableInfo = {
+    val isInternal = name.exists(_.startsWith(InternalAccumulator.METRICS_PREFIX))
+    new AccumulableInfo(id, name, update, value, isInternal, countFailedValues)
+  }
+
+  final private[spark] def isAtDriverSide: Boolean = atDriverSide
+
+  def isZero(): Boolean
```

Comment:
It's needed in https://github.com/apache/spark/pull/12612/files#diff-6a9ff7fb74fd490a50462d45db2d5e11R1101.

The logic is related to some UI stuff, I haven't looked into it yet, so I just keep it.